socketmode/socket_mode_managed_conn.go

```diff
@@ -0,0 +1,549 @@
+	upgradeHeader := http.Header{}
+	upgradeHeader.Add("Origin", "https://api.slack.com")
+	dialer := websocket.DefaultDialer
+	if smc.dialer != nil {
+		dialer = smc.dialer
+	}
+	conn, _, err := dialer.Dial(url, upgradeHeader)
+	if err != nil {
+		smc.Debugf("Failed to dial to the websocket: %s", err)
+		return nil, nil, err
+	}
+
+	conn.SetPingHandler(func(appData string) error {
+		if additionalPingHandler != nil {
+			if err := additionalPingHandler(appData); err != nil {
+				return err
+			}
+		}
+
+		smc.handlePing(conn, appData)
+
+		return nil
+	})
+
+	// We don't need to conn.SetCloseHandler because the default handler is effective enough that
+	// it sends back the CLOSE message to the server and let conn.ReadJSON() fail with CloseError.
+	// The CloseError must be handled normally in our receiveMessagesInto function.
+	//conn.SetCloseHandler(func(code int, text string) error {
+	//  ...
+	// })
+
+	return info, conn, err
+}
+
+// runResponseSender runs the handler that reads Socket Mode responses enqueued onto Client.socketModeResponses channel
+// and sends them one by one over the WebSocket connection.
+// Gorilla WebSocket is not goroutine safe hence this needs to be the single place you write to the WebSocket connection.
+func (smc *Client) runResponseSender(ctx context.Context, conn *websocket.Conn) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		// 3. listen for messages that need to be sent
+		case res := <-smc.socketModeResponses:
+			smc.Debugf("Sending Socket Mode response with envelope ID %q: %v", res.EnvelopeID, res)
+
+			if err := unsafeWriteSocketModeResponse(conn, res); err != nil {
+				smc.Events <- newEvent(EventTypeErrorWriteFailed, &ErrorWriteFailed{
+					Cause:    err,
+					Response: res,
+				})
+			}
+
+			smc.Debugf("Finished sending Socket Mode response with envelope ID %q", res.EnvelopeID)
+		}
+	}
+}
+
+// runRequestHandler is a blocking function that runs the Socket Mode request receiver.
+//
+// It reads WebSocket messages sent from Slack's Socket Mode WebSocket connection,
+// parses them as Socket Mode requests, and processes them and optionally emit our own events into Client.Events channel.
+func (smc *Client) runRequestHandler(ctx context.Context, websocket chan json.RawMessage) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case message := <-websocket:
+			smc.Debugf("Received WebSocket message: %s", message)
+
+			// listen for incoming messages that need to be parsed
+			evt, err := smc.parseEvent(message)
+			if err != nil {
+				smc.Events <- newEvent(EventTypeErrorBadMessage, &ErrorBadMessage{
+					Cause:   err,
+					Message: message,
+				})
+			} else if evt != nil {
+				if evt.Type == EventTypeDisconnect {
+					// We treat the `disconnect` request from Slack as an error internally,
+					// so that we can tell the consumer of this function to reopen the connection on it.
+					return ErrorRequestedDisconnect{}
```

Comment:
How does the handling code look like? Do developers need to write a handler on their own? If so, having the code in the example should be helpful. The disconnect requests can be regularly sent to all Socket Mode apps. This is not a rare case.

If this library handle the reconnection logic (= acquiring a new WSS URL and establishing a new connection to the endpoint under the hood), that's amazing!

Reply:
> If this library can have the reconnection logic (= acquiring a new WSS URL and establishing a new connection to the endpoint under the hood), that's amazing!

Good question - Yes I've made the client is able to handle the reconnection on its own, following all the awesome Socket Mode clients you've written for other languages 😄 

For your reference, the below is where we implement the infinite reconnection loop:

https://github.com/mumoshu/slack/blob/e9e7822646b5341295e0854167d774e4fb5cd33b/socketmode/socket_mode_managed_conn.go#L36-L42

Reply:
In go anything that is NOT under `internal` package and the name starts an uppercase later is considered public - so `ErrorRequestedDisconnect` is fairly confusing as it can give users misconception that one might want to use this outside of this package.

I'm going to make it private so that it is clear that the consumer of the client does not need to know about this internal error happens on disconnect request.

Reply:
>so ErrorRequestedDisconnect is fairly confusing as it can give users misconception that one might want to use this outside of this package.

Yes, actually I was thinking so too. Making it private should be the right direction 👍 

Also, I've verified the reconnecting logic works properly. Great work!